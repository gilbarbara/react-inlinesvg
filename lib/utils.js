'use strict';

exports.__esModule = true;
exports.configurationError = exports.unsupportedBrowserError = exports.uniquifySVGIDs = exports.randomString = exports.isSupportedEnvironment = exports.supportsInlineSVG = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _once = require('once');

var _once2 = _interopRequireDefault(_once);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const supportsInlineSVG = exports.supportsInlineSVG = (0, _once2.default)(() => {
  /* istanbul ignore next */
  if (!document) {
    return false;
  }

  const div = document.createElement('div');
  div.innerHTML = '<svg />';
  return div.firstChild && div.firstChild.namespaceURI === 'http://www.w3.org/2000/svg';
});

const isSupportedEnvironment = exports.isSupportedEnvironment = (0, _once2.default)(() => ((typeof window !== 'undefined' && window !== null ? window.XMLHttpRequest : false) || (typeof window !== 'undefined' && window !== null ? window.XDomainRequest : false)) && supportsInlineSVG());

const randomString = exports.randomString = (length = 8) => {
  const letters = 'abcdefghijklmnopqrstuvwxyz';
  const numbers = '1234567890';
  const charset = letters + letters.toUpperCase() + numbers;

  const randomCharacter = array => array[Math.floor(Math.random() * array.length)];

  let R = '';
  for (let i = 0; i < length; i++) {
    R += randomCharacter(charset);
  }
  return R;
};

const uniquifySVGIDs = exports.uniquifySVGIDs = (() => {
  const mkAttributePattern = attr => `(?:(?:\\s|\\:)${attr})`;

  const idPattern = new RegExp(`(?:(${mkAttributePattern('id')})="([^"]+)")|(?:(${mkAttributePattern('href')}|${mkAttributePattern('role')}|${mkAttributePattern('arcrole')})="\\#([^"]+)")|(?:="url\\(\\#([^\\)]+)\\)")`, 'g');

  return (svgText, svgID) => {
    const uniquifyID = id => `${id}___${svgID}`;

    return svgText.replace(idPattern, (m, p1, p2, p3, p4, p5) => {
      //eslint-disable-line consistent-return
      /* istanbul ignore else */
      if (p2) {
        return `${p1}="${uniquifyID(p2)}"`;
      } else if (p4) {
        return `${p3}="#${uniquifyID(p4)}"`;
      } else if (p5) {
        return `="url(#${uniquifyID(p5)})"`;
      }
    });
  };
})();

class InlineSVGError extends Error {
  constructor(message) {
    super();

    this.name = 'InlineSVGError';
    this.isSupportedBrowser = true;
    this.isConfigurationError = false;
    this.isUnsupportedBrowserError = false;
    this.message = message;

    return this;
  }
}

const createError = (message, attrs) => {
  const err = new InlineSVGError(message);

  return _extends({}, err, attrs);
};

const unsupportedBrowserError = exports.unsupportedBrowserError = message => {
  let newMessage = message;

  /* istanbul ignore else */
  if (!newMessage) {
    newMessage = 'Unsupported Browser';
  }

  return createError(newMessage, {
    isSupportedBrowser: false,
    isUnsupportedBrowserError: true
  });
};

const configurationError = exports.configurationError = message => createError(message, {
  isConfigurationError: true
});