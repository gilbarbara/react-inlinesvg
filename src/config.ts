export const CACHE_NAME = 'react-inlinesvg';
export const CACHE_MAX_RETRIES = 10;

export const STATUS = {
  IDLE: 'idle',
  LOADING: 'loading',
  LOADED: 'loaded',
  FAILED: 'failed',
  READY: 'ready',
  UNSUPPORTED: 'unsupported',
} as const;
