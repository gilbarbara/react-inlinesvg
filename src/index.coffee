# react-inlinesvg
# ===============

# Our dependencies.

React = require 'react'
once = require 'once'
httpplease = require 'httpplease'
ieXDomain = require 'httpplease/plugins/oldiexdomain'

{PropTypes} = React
{span} = React.DOM
http = httpplease.use ieXDomain

# An enum containing all the possible statuses of an isvg component. This is
# accessible via `InlineSVG.Status`.

Status =
  PENDING: 'pending'
  LOADING: 'loading'
  LOADED: 'loaded'
  FAILED: 'failed'
  UNSUPPORTED: 'unsupported'

# Utils
# -----
#
# A test to determine whether the browser supports inline SVGs.

supportsInlineSVG = once ->
  return false unless document
  div = document.createElement 'div'
  div.innerHTML = '<svg />'
  div.firstChild and div.firstChild.namespaceURI is 'http://www.w3.org/2000/svg'

# Wrap a function in a `setTimeout` call. This is used to guarantee async
# behavior, which can avoid unexpected errors.

delay = (fn) ->
  (args...) ->
    newFunc = -> fn args...
    setTimeout newFunc, 0
    return

# Our default support test.

isSupportedEnvironment = once ->
  (window?.XMLHttpRequest or window?.XDomainRequest) and supportsInlineSVG()

# Replaces the IDs in the provided string with document-unique IDs. This is
# obviously very naive, but is intended to provide a good-enough solution
# without adding too much overhead as would, for example, parsing as XML. See
# GH-3.

uniquifyIDs = do ->
  mkAttributePattern = (attr) -> "(?:(?:\\s|\\:)#{ attr })"
  idPattern =
    ///
    (?:                                    # Match ID declarations:
      (
        #{ mkAttributePattern 'id' }       #   The start of an attribute named "id"
      )
      ="
        ([^"]+)                            #   Our attribute value—match anything that isn't a quote.
      "                                    #   The end of the attribute
    )
    |                                      # ...OR...
    (?:                                    # Match IRI references:
      (
        #{ mkAttributePattern 'href' }
        |
        #{ mkAttributePattern 'role' }
        |
        #{ mkAttributePattern 'arcrole' }
      )
      ="                                   #   The start of an attribute
        \#                                 #   A literal # indicating this is an IRI reference
        ([^"]+)                            #   Our attribute value—match anything that isn't a quote.
      "                                    #   The end of the attribute
    )
    |                                      # ...OR...
    (?:                                    # Match FuncIRI references
      ="                                   #   The start of an attribute
        url\(                              #   The beginning of a FuncIRI
          \#                               #   A literal # indicating this is an IRI reference
          ([^\)]+)                         #   The id—match anything that isn't a closing paren
        \)                                 #   The end of the FuncIRI
      "                                    #   The end of the attribute
    )
    ///g
  (svgText, svgID) ->
    uniquifyID = (id) -> "#{ id }___#{ svgID }"
    svgText.replace idPattern, (m, p1, p2, p3, p4, p5) ->
      if p2 then "#{ p1 }=\"#{ uniquifyID p2 }\""
      else if p4 then "#{ p3 }=\"##{ uniquifyID p4 }\""
      else if p5 then "=\"url(##{ uniquifyID p5 })\""

# Generates a 32bit hash from the SVG path

getHash = (svgPath) ->
  hash = 0
  i = 0
  char = ''
  if svgPath.length == 0
    return hash
  while i < svgPath.length
    char = svgPath.charCodeAt(i)
    hash = (hash << 5) - hash + char
    hash = hash & hash
    i++
  hash

# Errors
# ------
#
# A custom error used by the InlineSVG component. For ease of inspection, we have
# one error type with several flags that can be used to get information about the
# error type.

class InlineSVGError extends Error
  name: 'InlineSVGError'
  isSupportedBrowser: true
  isConfigurationError: false
  isUnsupportedBrowserError: false
  constructor: (@message) ->

# Utility functions for creating `InlineSVGError` instances.

createError = (message, attrs) ->
  err = new InlineSVGError message
  for own k, v of attrs
    err[k] = v
  err

unsupportedBrowserError = (message = 'Unsupported Browser') ->
  createError message,
    isSupportedBrowser: false
    isUnsupportedBrowserError: true

configurationError = (message) ->
  createError message, isConfigurationError: true

# The InlineSVG component

module.exports = me =
  React.createClass
    statics: {Status}
    displayName: 'InlineSVG'
    propTypes:
      wrapper: PropTypes.func
      src: PropTypes.string.isRequired
      className: PropTypes.string
      preloader: PropTypes.func
      onLoad: PropTypes.func
      onError: PropTypes.func
      supportTest: PropTypes.func
      uniquifyIDs: PropTypes.bool
    getDefaultProps: ->
      wrapper: span
      supportTest: isSupportedEnvironment
      uniquifyIDs: true
    getInitialState: ->
      status: Status.PENDING
    componentDidMount: ->
      return unless @state.status is Status.PENDING

      # Either load the SVG or trigger an error, simulating async behavior with
      # `delay` so as not to violate any implicit expectations in the user's
      # code which would be difficult to troubleshoot.
      if @props.supportTest()
        if @props.src then @setState status: Status.LOADING, @load
        else do delay => @fail configurationError 'Missing source'
      else
        do delay => @fail unsupportedBrowserError()
    fail: (error) ->
      status = if error.isUnsupportedBrowserError then Status.UNSUPPORTED else Status.FAILED
      @setState {status}, => @props.onError? error
    handleLoad: (err, res) ->
      return @fail err if err

      # If the component has been unmounted since we started the load, just
      # forget it. (Setting the state of an unmounted component causes an
      # error.)
      return unless @isMounted()

      # Update the state to include the loaded text. This will be rendered to
      # the DOM the next time `render()` runs.
      @setState
        loadedText: res.text
        status: Status.LOADED
        => @props.onLoad?()
    load: -> http.get @props.src, @handleLoad
    getClassName: ->
      # Build a CSS class name based on the current state.
      className = "isvg #{ @state.status }"
      className += " #{ @props.className }" if @props.className
      className
    render: ->
      (@props.wrapper
        className: @getClassName()
        dangerouslySetInnerHTML: __html: @processSVG(@state.loadedText) if @state.loadedText
        @renderContents()
      )
    processSVG: (svgText) ->
      if @props.uniquifyIDs then uniquifyIDs svgText, getHash @props.src
      else svgText

    renderContents: ->
      switch @state.status
        when Status.UNSUPPORTED then @props.children
        when Status.PENDING, Status.LOADING
          new @props.preloader if @props.preloader
