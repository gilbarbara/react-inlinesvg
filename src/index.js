/* eslint-disable no-underscore-dangle */
import React from 'react';
import PropTypes from 'prop-types';
import DomToReact from 'dom-to-react';

import {
  canUseDOM,
  InlineSVGError,
  isSupportedEnvironment,
  randomString,
} from './utils';

const STATUS = {
  PENDING: 'pending',
  LOADING: 'loading',
  LOADED: 'loaded',
  FAILED: 'failed',
  READY: 'ready',
  UNSUPPORTED: 'unsupported'
};

export const storage = [];

export default class InlineSVG extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      content: '',
      element: null,
      hasCache: props.cacheRequests && !!storage.find(s => s.url === props.src),
      status: STATUS.PENDING
    };

    this._isMounted = false;
  }

  static propTypes = {
    baseURL: PropTypes.string,
    cacheRequests: PropTypes.bool,
    children: PropTypes.node,
    description: PropTypes.string,
    loader: PropTypes.node,
    onError: PropTypes.func,
    onLoad: PropTypes.func,
    preProcessor: PropTypes.func,
    src: PropTypes.string.isRequired,
    title: PropTypes.string,
    uniqueHash: PropTypes.string,
    uniquifyIDs: PropTypes.bool
  };

  static defaultProps = {
    cacheRequests: true,
    onLoad: () => {},
    uniquifyIDs: false
  };

  componentDidMount() {
    this._isMounted = true;

    if (!canUseDOM()) {
      this.handleError(new InlineSVGError('No DOM'));
      return;
    }

    const { status } = this.state;
    const { src } = this.props;

    try {
      /* istanbul ignore else */
      if (status === STATUS.PENDING) {
        /* istanbul ignore else */
        if (!isSupportedEnvironment()) throw new InlineSVGError('Browser does not support SVG');

        /* istanbul ignore else */
        if (!src) throw new InlineSVGError('Missing src');

        this.load();
      }
    }

    catch (error) {
      this.handleError(error);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (!canUseDOM()) return;

    const { hasCache, status } = this.state;
    const { onLoad, src } = this.props;

    if (prevState.status !== STATUS.READY && status === STATUS.READY) {
      onLoad(src, hasCache);
    }

    if (prevProps.src !== src) {
      if (!src) {
        this.handleError(new InlineSVGError('Missing src'));
        return;
      }

      this.load();
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  parseSVG() {
    const { content } = this.state;
    const { preProcessor } = this.props;

    if (preProcessor) {
      return preProcessor(content);
    }

    return content;
  }

  updateSVGAttributes(node) {
    const { baseURL = '', uniquifyIDs, uniqueHash } = this.props;
    const replaceableAttributes = [
      'id',
      'href',
      'xlink:href',
      'xlink:role',
      'xlink:arcrole'
    ];

    if (!uniquifyIDs) {
      return node;
    }

    const hash = uniqueHash || randomString();

    [...node.childNodes].forEach(d => {
      if (d.attributes && d.attributes.length) {
        const attributes = Object.values(d.attributes);

        attributes.forEach(a => {
          const match = a.value.match(/^url\((#[^)]+)/);

          if (match && match[1]) {
            a.value = `url(${baseURL}${match[1]}__${hash})`;
          }
        });

        replaceableAttributes.forEach(r => {
          const attribute = attributes.find(a => a.name === r);

          if (attribute) {
            attribute.value = `${attribute.value}__${hash}`;
          }
        });
      }

      if (d.childNodes.length) {
        d = this.updateSVGAttributes(d); // eslint-disable-line no-param-reassign
      }
    });

    return node;
  }

  generateNode() {
    const { description, title } = this.props;

    try {
      const svgText = this.parseSVG();
      const parser = new DOMParser();
      const doc = parser.parseFromString(svgText, 'image/svg+xml');
      const svg = this.updateSVGAttributes(doc.querySelector('svg'));

      if (description) {
        const originalDesc = svg.querySelector('desc');

        if (originalDesc) {
          originalDesc.innerHTML = description;
        }
        else {
          const descElement = document.createElement('desc');
          descElement.innerHTML = description;
          svg.prepend(descElement);
        }
      }

      if (title) {
        const originalTitle = svg.querySelector('title');

        if (originalTitle) {
          originalTitle.innerHTML = title;
        }
        else {
          const titleElement = document.createElement('title');
          titleElement.innerHTML = title;
          svg.prepend(titleElement);
        }
      }

      return svg;
    }
    catch (error) {
      return this.handleError(error);
    }
  }

  generateElement() {
    const {
      baseURL,
      cacheRequests,
      children,
      description,
      onError,
      onLoad,
      loader,
      preProcessor,
      src,
      title,
      uniqueHash,
      uniquifyIDs,
      ...rest
    } = this.props;

    try {
      const node = this.generateNode();

      /* istanbul ignore else */
      if (node) {
        const d2r = new DomToReact();

        this.setState({
          element: React.cloneElement(d2r.prepareNode(node), { ...rest }),
          status: STATUS.READY
        });
      }
    }
    catch (error) {
      this.handleError(error);
    }
  }

  load() {
    /* istanbul ignore else */
    if (this._isMounted) {
      this.setState(
        {
          content: '',
          element: null,
          status: STATUS.LOADING
        },
        () => {
          const { cacheRequests, src } = this.props;
          const cache = cacheRequests && storage.find(d => d.url === src);

          if (cache) {
            this.handleLoad(cache.content);
            return;
          }

          const dataURI = src.match(/data:image\/svg[^,]*?(;base64)?,(.*)/);
          let inlineSrc;

          if (dataURI) {
            inlineSrc = dataURI[1]
              ? atob(dataURI[2])
              : decodeURIComponent(dataURI[2]);
          }
          else if (src.indexOf('<svg') >= 0) {
            inlineSrc = src;
          }

          if (inlineSrc) {
            this.handleLoad(inlineSrc);
            return;
          }

          this.request();
        }
      );
    }
  }

  handleLoad = content => {
    /* istanbul ignore else */
    if (this._isMounted) {
      this.setState(
        {
          content,
          status: STATUS.LOADED
        },
        this.generateElement
      );
    }
  };

  handleError = error => {
    const { onError } = this.props;
    const status = error.message === 'Browser does not support SVG'
      ? STATUS.UNSUPPORTED
      : STATUS.FAILED;

    /* istanbul ignore else */
    if (this._isMounted) {
      this.setState({ status }, () => {
        /* istanbul ignore else */
        if (typeof onError === 'function') {
          onError(error);
        }
      });
    }
  };

  request = () => {
    const { cacheRequests, src } = this.props;

    return fetch(src)
      .then(response => {
        if (response.status > 299) {
          throw new InlineSVGError('request: Not found');
        }

        return response.text();
      })
      .then(content => {
        /* istanbul ignore else */
        if (cacheRequests) {
          storage.push({ url: src, content });
        }

        this.handleLoad(content);
      })
      .catch(error => this.handleError(error));
  };

  render() {
    if (!canUseDOM()) return null;

    const { element, status } = this.state;
    const { children = null, loader = null } = this.props;

    if (element) {
      return element;
    }

    if ([STATUS.UNSUPPORTED, STATUS.FAILED].includes(status)) {
      return children;
    }

    return loader;
  }
}
