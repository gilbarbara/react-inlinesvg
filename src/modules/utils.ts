import convert from 'react-from-dom';

import { Props, State } from '../types';

interface GetNodeOptions extends Props, Pick<State, 'content'> {
  handleError: (error: Error) => void;
  hash: string;
}

interface UpdateSVGAttributesOptions extends Pick<Props, 'baseURL' | 'uniquifyIDs'> {
  hash: string;
}

export function getNode(options: GetNodeOptions) {
  const {
    baseURL,
    content,
    description,
    handleError,
    hash,
    preProcessor,
    title,
    uniquifyIDs = false,
  } = options;

  try {
    const svgText = processSVG(content, preProcessor);
    const node = convert(svgText, { nodeOnly: true });

    if (!node || !(node instanceof SVGSVGElement)) {
      throw new Error('Could not convert the src to a DOM Node');
    }

    const svg = updateSVGAttributes(node, { baseURL, hash, uniquifyIDs });

    if (description) {
      const originalDesc = svg.querySelector('desc');

      if (originalDesc?.parentNode) {
        originalDesc.parentNode.removeChild(originalDesc);
      }

      const descElement = document.createElementNS('http://www.w3.org/2000/svg', 'desc');

      descElement.innerHTML = description;
      svg.prepend(descElement);
    }

    if (typeof title !== 'undefined') {
      const originalTitle = svg.querySelector('title');

      if (originalTitle?.parentNode) {
        originalTitle.parentNode.removeChild(originalTitle);
      }

      if (title) {
        const titleElement = document.createElementNS('http://www.w3.org/2000/svg', 'title');

        titleElement.innerHTML = title;
        svg.prepend(titleElement);
      }
    }

    return svg;
  } catch (error: any) {
    return handleError(error);
  }
}

export function processSVG(content: string, preProcessor?: Props['preProcessor']) {
  if (preProcessor) {
    return preProcessor(content);
  }

  return content;
}

export function updateSVGAttributes(
  node: SVGSVGElement,
  options: UpdateSVGAttributesOptions,
): SVGSVGElement {
  const { baseURL = '', hash, uniquifyIDs } = options;
  const replaceableAttributes = ['id', 'href', 'xlink:href', 'xlink:role', 'xlink:arcrole'];
  const linkAttributes = ['href', 'xlink:href'];
  const isDataValue = (name: string, value: string) =>
    linkAttributes.includes(name) && (value ? !value.includes('#') : false);

  if (!uniquifyIDs) {
    return node;
  }

  [...node.children].forEach(d => {
    if (d.attributes?.length) {
      const attributes = Object.values(d.attributes).map(a => {
        const attribute = a;
        const match = /url\((.*?)\)/.exec(a.value);

        if (match?.[1]) {
          attribute.value = a.value.replace(match[0], `url(${baseURL}${match[1]}__${hash})`);
        }

        return attribute;
      });

      replaceableAttributes.forEach(r => {
        const attribute = attributes.find(a => a.name === r);

        if (attribute && !isDataValue(r, attribute.value)) {
          attribute.value = `${attribute.value}__${hash}`;
        }
      });
    }

    if (d.children.length) {
      return updateSVGAttributes(d as SVGSVGElement, options);
    }

    return d;
  });

  return node;
}
