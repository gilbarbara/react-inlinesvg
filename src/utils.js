import once from 'once';

export const supportsInlineSVG = once(() => {
  /* istanbul ignore next */
  if (!document) {
    return false;
  }

  const div = document.createElement('div');
  div.innerHTML = '<svg />';
  return div.firstChild && div.firstChild.namespaceURI === 'http://www.w3.org/2000/svg';
});

export const isSupportedEnvironment = once(() => (
  supportsInlineSVG()
    && typeof window !== 'undefined' && window !== null
    ? window.XMLHttpRequest || window.XDomainRequest
    : false
));

export const randomString = (length = 8) => {
  const letters = 'abcdefghijklmnopqrstuvwxyz';
  const numbers = '1234567890';
  const charset = letters + letters.toUpperCase() + numbers;

  const randomCharacter = array => array[Math.floor(Math.random() * array.length)];

  let R = '';
  for (let i = 0; i < length; i++) {
    R += randomCharacter(charset);
  }
  return R;
};

export const uniquifySVGIDs = (() => {
  const mkAttributePattern = attr => `(?:(?:\\s|\\:)${attr})`;

  const idPattern = new RegExp(`(?:(${(mkAttributePattern('id'))})="([^"]+)")|(?:(${(mkAttributePattern('href'))}|${(mkAttributePattern('role'))}|${(mkAttributePattern('arcrole'))})="\\#([^"]+)")|(?:="url\\(\\#([^\\)]+)\\)")|(?:url\\(\\#([^\\)]+)\\))`, 'g');

  return (svgText, svgID, baseURL) => {
    const uniquifyID = id => `${id}___${svgID}`;

    return svgText.replace(idPattern, (m, p1, p2, p3, p4, p5, p6) => { //eslint-disable-line consistent-return
      /* istanbul ignore else */
      if (p2) {
        return `${p1}="${(uniquifyID(p2))}"`;
      }

      if (p4) {
        return `${p3}="${baseURL}#${(uniquifyID(p4))}"`;
      }

      if (p5) {
        return `="url(${baseURL}#${(uniquifyID(p5))})"`;
      }

      if (p6) {
        return `url(${baseURL}#${uniquifyID(p6)})`;
      }
    });
  };
})();

class InlineSVGError extends Error {
  constructor(message) {
    super();

    this.name = 'InlineSVGError';
    this.isSupportedBrowser = true;
    this.isConfigurationError = false;
    this.isUnsupportedBrowserError = false;
    this.message = message;

    return this;
  }
}

const createError = (message, attrs) => {
  const err = new InlineSVGError(message);

  return {
    ...err,
    ...attrs,
  };
};

export const unsupportedBrowserError = message => {
  let newMessage = message;

  /* istanbul ignore else */
  if (!newMessage) {
    newMessage = 'Unsupported Browser';
  }

  return createError(newMessage, {
    isSupportedBrowser: false,
    isUnsupportedBrowserError: true
  });
};

export const configurationError = message => createError(message, {
  isConfigurationError: true
});
