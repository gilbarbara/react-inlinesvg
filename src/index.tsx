/* eslint-disable no-underscore-dangle */
import React from 'react';

// @ts-ignore
import DomToReact from 'dom-to-react';

import { canUseDOM, InlineSVGError, isSupportedEnvironment, randomString } from './utils';

export interface Props {
  baseURL?: string;
  cacheRequests?: boolean;
  children?: React.ReactNode;
  description?: string;
  loader?: React.ReactNode;
  onError?: (error: InlineSVGError | FetchError) => void;
  onLoad?: (src: URL | string, isCached: boolean) => void;
  preProcessor?: (code: string) => string;
  src: string;
  title?: string;
  uniqueHash?: string;
  uniquifyIDs?: boolean;
  [key: string]: any;
}

export interface State {
  content: string;
  element: React.ReactNode;
  hasCache: boolean;
  status: string;
}

export interface FetchError extends Error {
  message: string;
  type: string;
  errno: string;
  code: string;
}

export interface StorageItem {
  url: string;
  content: string;
}

export const STATUS = {
  FAILED: 'failed',
  LOADED: 'loaded',
  LOADING: 'loading',
  PENDING: 'pending',
  READY: 'ready',
  UNSUPPORTED: 'unsupported',
};

export const storage: StorageItem[] = [];

export default class InlineSVG extends React.PureComponent<Props, State> {
  private static defaultProps = {
    cacheRequests: true,
    uniquifyIDs: false,
  };

  // tslint:disable-next-line:variable-name
  private _isMounted: boolean;

  constructor(props: Props) {
    super(props);

    this.state = {
      content: '',
      element: null,
      hasCache: !!props.cacheRequests && !!storage.find((s: StorageItem) => s.url === props.src),
      status: STATUS.PENDING,
    };

    this._isMounted = false;
  }

  public componentDidMount() {
    this.isMounted = true;

    if (!canUseDOM()) {
      this.handleError(new InlineSVGError('No DOM'));
      return;
    }

    const { status } = this.state;
    const { src } = this.props;

    try {
      /* istanbul ignore else */
      if (status === STATUS.PENDING) {
        /* istanbul ignore else */
        if (!isSupportedEnvironment()) {
          throw new InlineSVGError('Browser does not support SVG');
        }

        /* istanbul ignore else */
        if (!src) {
          throw new InlineSVGError('Missing src');
        }

        this.load();
      }
    } catch (error) {
      this.handleError(error);
    }
  }

  public componentDidUpdate(prevProps: Props, prevState: State) {
    if (!canUseDOM()) {
      return;
    }

    const { hasCache, status } = this.state;
    const { onLoad, src } = this.props;

    if (prevState.status !== STATUS.READY && status === STATUS.READY) {
      /* istanbul ignore else */
      if (onLoad) {
        onLoad(src, hasCache);
      }
    }

    if (prevProps.src !== src) {
      if (!src) {
        this.handleError(new InlineSVGError('Missing src'));
        return;
      }

      this.load();
    }
  }

  public componentWillUnmount() {
    this._isMounted = false;
  }

  get isMounted() {
    return this._isMounted;
  }

  set isMounted(value: boolean) {
    this._isMounted = value;
  }

  private parseSVG() {
    const { content } = this.state;
    const { preProcessor } = this.props;

    if (preProcessor) {
      return preProcessor(content);
    }

    return content;
  }

  private updateSVGAttributes(node: SVGSVGElement): SVGSVGElement {
    const { baseURL = '', uniquifyIDs, uniqueHash } = this.props;
    const replaceableAttributes = ['id', 'href', 'xlink:href', 'xlink:role', 'xlink:arcrole'];

    if (!uniquifyIDs) {
      return node;
    }

    const hash = uniqueHash || randomString();

    [...node.children].forEach(d => {
      if (d.attributes && d.attributes.length) {
        const attributes = Object.values(d.attributes);

        attributes.forEach(a => {
          const match = a.value.match(/^url\((#[^)]+)/);

          if (match && match[1]) {
            a.value = `url(${baseURL}${match[1]}__${hash})`;
          }
        });

        replaceableAttributes.forEach(r => {
          const attribute = attributes.find(a => a.name === r);

          if (attribute) {
            attribute.value = `${attribute.value}__${hash}`;
          }
        });
      }

      if (d.children.length) {
        d = this.updateSVGAttributes(d as SVGSVGElement);
      }
    });

    return node;
  }

  private generateNode() {
    const { description, title } = this.props;

    try {
      const svgText = this.parseSVG();
      const parser = new DOMParser();
      const doc = parser.parseFromString(svgText, 'image/svg+xml');
      let svg = doc.querySelector('svg');

      if (!svg) {
        throw new InlineSVGError('Could not parse the SVG code');
      }

      svg = this.updateSVGAttributes(svg);

      if (description) {
        const originalDesc = svg.querySelector('desc');

        if (originalDesc && originalDesc.parentNode) {
          originalDesc.parentNode.removeChild(originalDesc);
        }

        const descElement = document.createElement('desc');
        descElement.innerHTML = description;
        svg.prepend(descElement);
      }

      if (title) {
        const originalTitle = svg.querySelector('title');

        if (originalTitle && originalTitle.parentNode) {
          originalTitle.parentNode.removeChild(originalTitle);
        }

        const titleElement = document.createElement('title');
        titleElement.innerHTML = title;
        svg.prepend(titleElement);
      }

      return svg;
    } catch (error) {
      return this.handleError(error);
    }
  }

  private generateElement() {
    const {
      baseURL,
      cacheRequests,
      children,
      description,
      onError,
      onLoad,
      loader,
      preProcessor,
      src,
      title,
      uniqueHash,
      uniquifyIDs,
      ...rest
    } = this.props;

    try {
      const node = this.generateNode();

      /* istanbul ignore else */
      if (node) {
        const d2r = new DomToReact();

        this.setState({
          element: React.cloneElement(d2r.prepareNode(node), { ...rest }),
          status: STATUS.READY,
        });
      }
    } catch (error) {
      this.handleError(error);
    }
  }

  private load() {
    /* istanbul ignore else */
    if (this.isMounted) {
      this.setState(
        {
          content: '',
          element: null,
          status: STATUS.LOADING,
        },
        () => {
          const { cacheRequests, src } = this.props;
          const cache = cacheRequests && storage.find(d => d.url === src);

          if (cache) {
            this.handleLoad(cache.content);
            return;
          }

          const dataURI = src.match(/data:image\/svg[^,]*?(;base64)?,(.*)/);
          let inlineSrc;

          if (dataURI) {
            inlineSrc = dataURI[1] ? atob(dataURI[2]) : decodeURIComponent(dataURI[2]);
          } else if (src.indexOf('<svg') >= 0) {
            inlineSrc = src;
          }

          if (inlineSrc) {
            this.handleLoad(inlineSrc);
            return;
          }

          this.request();
        },
      );
    }
  }

  private handleLoad = (content: string) => {
    /* istanbul ignore else */
    if (this.isMounted) {
      this.setState(
        {
          content,
          status: STATUS.LOADED,
        },
        this.generateElement,
      );
    }
  };

  private handleError = (error: InlineSVGError | FetchError) => {
    const { onError } = this.props;
    const status =
      error.message === 'Browser does not support SVG' ? STATUS.UNSUPPORTED : STATUS.FAILED;

    if (process.env.NODE_ENV === 'development') {
      console.error(error); // tslint:disable-line:no-console
    }

    /* istanbul ignore else */
    if (this.isMounted) {
      this.setState({ status }, () => {
        /* istanbul ignore else */
        if (typeof onError === 'function') {
          onError(error);
        }
      });
    }
  };

  private request = () => {
    const { cacheRequests, src } = this.props;

    try {
      return fetch(src)
        .then(response => {
          if (response.status > 299) {
            throw new InlineSVGError('Not Found');
          }

          return response.text();
        })
        .then(content => {
          /* istanbul ignore else */
          if (cacheRequests) {
            storage.push({ url: src, content });
          }

          this.handleLoad(content);
        })
        .catch(error => this.handleError(error));
    } catch (error) {
      this.handleError(new InlineSVGError(error.message));
    }
  };

  public render() {
    if (!canUseDOM()) {
      return null;
    }

    const { element, status } = this.state;
    const { children = null, loader = null } = this.props;

    if (element) {
      return element;
    }

    if ([STATUS.UNSUPPORTED, STATUS.FAILED].indexOf(status) > -1) {
      return children;
    }

    return loader;
  }
}
