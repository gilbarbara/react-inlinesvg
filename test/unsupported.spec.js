import React from 'react';
import { InlineSVGError } from '../src/utils';

const fixtures = {
  play: 'http://localhost:1337/play.svg',
};

const mockOnError = jest.fn();
let mockCanUseDOM = false;
let mockIsSupportedEnvironment = true;

jest.mock('../src/utils', () => {
  const utils = require.requireActual('../src/utils');

  return {
    ...utils,
    canUseDOM: () => mockCanUseDOM,
    isSupportedEnvironment: () => mockIsSupportedEnvironment,
  };
});

function setup({ onLoad, ...rest }) {
  // eslint-disable-next-line global-require
  const ReactInlineSVG = require('../src').default;

  return new Promise(resolve => {
    const wrapper = mount(<ReactInlineSVG
      onLoad={(src, isCached) => {
        setTimeout(() => {
          if (onLoad) onLoad(src, isCached);

          resolve(wrapper);
        }, 0);
      }}
      onError={error => {
        mockOnError(error);
        setTimeout(() => resolve(wrapper), 0);
      }}
      {...rest}
    />);

    return wrapper;
  });
}

describe('unsupported environments', () => {
  it('shouldn\'t break without DOM ', async() => {
    const wrapper = await setup({
      src: fixtures.play,
    });

    expect(mockOnError).toHaveBeenCalledWith(new InlineSVGError('No DOM'));
    expect(wrapper).toMatchSnapshot();

    //  expect(wrapper.state('content')).toBe(fixtures.play);
  });

  xit('should warn the user if fetch is missing', () => {

  });

  it('shouldn\'t not render anything if is an unsupported browser', async() => {
    mockCanUseDOM = true;
    mockIsSupportedEnvironment = false;

    const wrapper = await setup({
      src: fixtures.play,
    });

    expect(mockOnError).toHaveBeenCalledWith(new InlineSVGError('Browser does not support SVG'));
    expect(wrapper).toMatchSnapshot();
  });
});
