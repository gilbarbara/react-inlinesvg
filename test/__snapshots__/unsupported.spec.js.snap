// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unsupported environments should warn the user if fetch is missing 1`] = `
<InlineSVG
  cacheRequests={true}
  onError={[Function]}
  onLoad={[Function]}
  src="http://localhost:1337/play.svg"
  uniquifyIDs={false}
/>
`;

exports[`unsupported environments shouldn't break without DOM  1`] = `
<InlineSVG
  cacheRequests={true}
  onError={[Function]}
  onLoad={[Function]}
  src="http://localhost:1337/play.svg"
  uniquifyIDs={false}
/>
`;

exports[`unsupported environments shouldn't not render anything if is an unsupported browser 1`] = `
<InlineSVG
  cacheRequests={true}
  onError={[Function]}
  onLoad={[Function]}
  src="http://localhost:1337/play.svg"
  uniquifyIDs={false}
/>
`;
